R/fair-aaa.R

```diff
@@ -0,0 +1,181 @@
+#' output is itself a function. Further, the functions that this function
+#' outputs are also function factories. More explicitly, this looks like:
+#'
+#' ```
+#' # a function with similar implementation to `demographic_parity()`:
+#' diff_range <- function(x, ...) {diff(range(x$.estimate))}
+#'
+#' dem_parity <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "dem_parity",
+#'     .post = diff_range
+#'   )
+#' ```
+#'
+#' The outputted `dem_parity` is a function that takes one argument, `by`,
+#' indicating the data-masked variable giving the sensitive feature.
+#'
+#' When called with a `by` argument, `dem_parity` will return a yardstick
+#' metric function like any other:
+#'
+#' ```
+#' dem_parity_by_gender <- dem_parity(gender)
+#' ```
+#'
+#' Note that `dem_parity` doesn't take any arguments other than `by`, and thus
+#' knows nothing about the data it will be applied to other than that it ought
+#' to have a column with name `"gender"` in it.
+#'
+#' The output `dem_parity_by_gender` is a metric function that takes the
+#' same arguments as the function supplied as `fn`, in this case
+#' `detection_prevalence`. It will thus interface like any other yardstick
+#' function except that it will look for a `"gender"` column in
+#' the data it's supplied.
+#'
+#' In addition to the examples below, see the documentation on the
+#' return value of fairness metrics like [demographic_parity()],
+#' [equal_opportunity()], or [equalized_odds()] to learn more about how the
+#' output of this function can be used.
+#'
+#' @examples
+#' data(hpc_cv)
+#'
+#' # `demographic_parity`, among other fairness metrics,
+#' # is generated with `fairness_metric()`:
+#' diff_range <- function(x, ...) {diff(range(x$.estimate))}
+#' demographic_parity_ <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity",
+#'     .post = diff_range
+#'   )
+#'
+#' m_set <- metric_set(demographic_parity_(Resample))
+#'
+#' m_set(hpc_cv, truth = obs, estimate = pred)
+#'
+#' # the `post` argument can be used to accommodate a wide
+#' # variety of parameterizations. to encode demographic
+#' # parity as a ratio inside of a difference, for example:
+#' ratio_range <- function(x, ...) {
+#'   range <- range(x$.estimate)
+#'   range[1] / range[2]
+#' }
+#'
+#' demographic_parity_ratio <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity_ratio",
+#'     .post = ratio_range
+#'   )
+#'
+#' @export
+fairness_metric <- function(.fn, .name, .post) {
+  if (is_missing(.fn) || !inherits_any(.fn, c("metric", "metric_set"))) {
+    abort("`.fn` must be a metric function or metric set.")
+  }
+  if (is_missing(.name) || !is_string(.name)) {
+    abort("`.name` must be a string.")
+  }
+  if (is_missing(.post) || !is_function(.post)) {
+    abort("`.post` must be a function.")
+  }
+
+  function(by) {
+    by_str <- as_string(enexpr(by))
+    res <-
+      function(data, ...) {
```

Comment:
The `...` in this argument are passed to both `.fn` and to `.post`. We could:

* Remove the `...` argument to `.post`, since defining a problem-context-aware function means that all of the information you need likely lives in the metric set or can be hard-coded into `.post`
* Resolve `formals(.fn)` as the formals of this function, which would fully address the case where `.fn` is one yardstick function like `sens` or `detection_prevalence`; those functions don't use their dots, and `.post` would only get the dots it needs (i.e. doesn't need to know what to do with `truth` or `estimate`)
* Add a check a few lines above here to ensure that `.post` has a `...` in that argument. In that case, `truth`/`estimate`/etc are not `force()`d and users shouldn't run into issues as long as they don't define arguments to `.post` with the same name as an argument to `.fn`.

Reply:
See https://github.com/tidymodels/yardstick/actions/runs/4996847332/jobs/8950515558#step:6:178 for an example of how this comes up.

Reply:
Addressed in [45f57ff](https://github.com/tidymodels/yardstick/pull/434/commits/45f57ff6a1480a53d66ed1bbfe8cbba4d7b5762d) via the first bulleted option.